Mathlib/Combinatorics/SetFamily/KruskalKatona.lean

```diff
@@ -0,0 +1,448 @@
+difference. -/
+lemma compression_improved (𝒜 : Finset (Finset α)) (h₁ : UsefulCompression U V)
+    (h₂ : ∀ ⦃U₁ V₁⦄, UsefulCompression U₁ V₁ → U₁.card < U.card → IsCompressed U₁ V₁ 𝒜) :
+    (∂ (𝓒 U V 𝒜)).card ≤ (∂ 𝒜).card := by
+  obtain ⟨UVd, same_size, hU, hV, max_lt⟩ := h₁
+  refine card_shadow_compression_le _ _ fun x Hx ↦ ⟨min' V hV, min'_mem _ _, ?_⟩
+  obtain hU' | hU' := eq_or_lt_of_le (succ_le_iff.2 hU.card_pos)
+  · rw [← hU'] at same_size
+    have : erase U x = ∅ := by rw [← Finset.card_eq_zero, card_erase_of_mem Hx, ← hU']
+    have : erase V (min' V hV) = ∅ := by
+      rw [← Finset.card_eq_zero, card_erase_of_mem (min'_mem _ _), ← same_size]
+    rw [‹erase U x = ∅›, ‹erase V (min' V hV) = ∅›]
+    exact isCompressed_self _ _
+  refine h₂ ⟨UVd.mono (erase_subset ..) (erase_subset ..), ?_, ?_, ?_, ?_⟩ (card_erase_lt_of_mem Hx)
+  · rw [card_erase_of_mem (min'_mem _ _), card_erase_of_mem Hx, same_size]
+  · rwa [← card_pos, card_erase_of_mem Hx, tsub_pos_iff_lt]
+  · rwa [← Finset.card_pos, card_erase_of_mem (min'_mem _ _), ← same_size, tsub_pos_iff_lt]
+  · exact (Finset.max'_subset _ $ erase_subset _ _).trans_lt (max_lt.trans_le $
+      le_max' _ _ $ mem_erase.2 ⟨(min'_lt_max'_of_card _ (by rwa [← same_size])).ne', max'_mem _ _⟩)
+
+/-- If we're compressed by all useful compressions, then we're an initial segment. This is the other
+key Kruskal-Katona part. -/
+lemma isInitSeg_of_compressed {ℬ : Finset (Finset α)} {r : ℕ} (h₁ : (ℬ : Set (Finset α)).Sized r)
+    (h₂ : ∀ U V, UsefulCompression U V → IsCompressed U V ℬ) : IsInitSeg ℬ r := by
+  refine ⟨h₁, ?_⟩
+  rintro A B hA ⟨hBA, sizeA⟩
+  by_contra hB
+  have hAB : A ≠ B := ne_of_mem_of_not_mem hA hB
+  have hAB' : A.card = B.card := (h₁ hA).trans sizeA.symm
+  have hU : (A \ B).Nonempty := sdiff_nonempty.2 fun h ↦ hAB $ eq_of_subset_of_card_le h hAB'.ge
+  have hV : (B \ A).Nonempty :=
+    sdiff_nonempty.2 fun h ↦ hAB.symm $ eq_of_subset_of_card_le h hAB'.le
+  have disj : Disjoint (B \ A) (A \ B) := disjoint_sdiff.mono_left sdiff_subset
+  have smaller : max' _ hV < max' _ hU := by
+    obtain hlt | heq | hgt := lt_trichotomy (max' _ hU) (max' _ hV)
+    · rw [← compress_sdiff_sdiff A B] at hAB hBA
+      cases hBA.not_lt $ toColex_compress_lt_toColex hlt hAB
+    · exact (disjoint_right.1 disj (max'_mem _ hU) $ heq.symm ▸ max'_mem _ _).elim
+    · assumption
+  refine hB ?_
+  rw [← (h₂ _ _ ⟨disj, card_sdiff_comm hAB'.symm, hV, hU, smaller⟩).eq]
+  exact mem_compression.2 (Or.inr ⟨hB, A, hA, compress_sdiff_sdiff _ _⟩)
+
+attribute [-instance] Fintype.decidableForallFintype
+
+-- TODO: There's currently a diamond
+-- import Mathlib.Data.Fin.Basic
+-- example (n : ℕ) : instDecidableEqFin n = instDecidableEq := rfl
+attribute [-instance] instDecidableEqFin
+
+/-- This measures roughly how compressed the family is. (Note that it does depend on the order of
+the ground set, unlike Kruskal-Katona itself). -/
```

Comment:
This isn't strictly true, since the definition of initial segments of colex that's used below depends on the order of the ground set. What is true, however, is that the *size* of C doesn't depend on the order of the ground set. 
The original motivation for the todo on the order iso between colex and Nat was to make the statement of KK genuinely not depend on the order of the ground set, so I wonder if @apnelson1's recent work makes the statements below nicer.

Reply:
I mean, the better way of doing things is the k-cascade representation, right? It's a bit annoying that I lost my work on that due to some git misshap, but I could possibly redo it if you think that making the statement order-independent is so important

Reply:
I think we shouldn't write something in a docstring that's false

Reply:
I don't think the docstring is false? It's true that `kruskal_katona` is order-dependent, but it's also true that the (usual) mathematical statement of Kruskal-Katona is not.

Reply:
Then let's make sure that's clear in the docstring